### PR TITLE
Upgrade CAF to 1.0.2

### PR DESCRIPTION
**Pull request by tobim**
This is the real update PR. Functionally nothing changes, but several unit tests had to be removed because they relied testing functionality that does not exist any more.

The diff is still quite large, but those changes that could still be pulled out are smallish fixes that are unlikely to cause conflicts.

This should probably be directly after the next minor version release.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -21,7 +21,7 @@
 	shallow = true
 [submodule "libtenzir/aux/caf"]
 	path = libtenzir/aux/caf
-	url = https://github.com/tenzir/actor-framework
+  url = https://github.com/actor-framework/actor-framework
```
Comment (tobim) [inline .gitmodules:24]: Go back to the tab for indentation.

File: libtenzir/src/default_configuration.cpp [hunk 2]
```diff
@@ -21,38 +18,11 @@ default_configuration::default_configuration() {
...
-  //     make progress and the deadlock can be resolved.
-  // We do a combination of (1) and (2).
-  set("caf.stream.max-batch-delay", caf::timespan{15ms});
+      caf::make_config_value_list("caf", "caf_flow"));
+  // TODO: Validate these.
   set("caf.scheduler.max-throughput", 500);
-  set("caf.stream.credit-policy", "token-based");
-  set("caf.stream.token-based-policy.batch-size", 1);
-  set("caf.stream.token-based-policy.buffer-size", 64);
   set("caf.middleman.connection-timeout", caf::timespan{120s});
```
Comment (dominiklohmann) [inline libtenzir/src/default_configuration.cpp:26]: Did you validate these?

File: libtenzir/src/detail/installdirs_relocatable.cpp.in [hunk 3]
```diff
@@ -7,6 +7,7 @@
 // SPDX-License-Identifier: BSD-3-Clause
 
 #include "tenzir/detail/assert.hpp"
+#include <string_view>
```
Comment (dominiklohmann) [inline libtenzir/src/detail/installdirs_relocatable.cpp.in:10]: The include order looks wrong here. Likely not picked up by clang-format because it's a template file for CMake?

File: libtenzir/src/execution_node.cpp [hunk 4]
```diff
@@ -686,7 +687,10 @@ struct exec_node_state {
           TENZIR_DEBUG("{} {} failed to get its demand fulfilled: {}", *self,
                        op->name(), err);
           issue_demand_inflight = false;
-          if (err != caf::sec::request_receiver_down) {
+          if (err
+              and err != caf::sec::request_receiver_down
+              // TODO: verify that this is fine.
+              and err != caf::exit_reason::remote_link_unreachable) {
```
Comment (dominiklohmann) [inline libtenzir/src/execution_node.cpp:693]: This is fine if my understanding is correct. Which of the two error codes you get is no longer deterministic for remote actors. That's a bit unfortunate, but whatever.

File: libtenzir/test/flatbuffer.cpp [hunk 5]
```diff
@@ -50,37 +48,4 @@ TEST(lifetime) {
...
-    fbt = std::move(*maybe_fbt);
-    auto fbt2 = roundtrip(fbt);
-    CHECK_EQUAL(as_bytes(fbt.chunk()), as_bytes(fbt2.chunk()));
-  }
-  auto fbrt = fbt.slice(*fbt->type_as_record_type());
-  auto fbrtf = fbrt.slice(*fbrt->fields()->Get(0));
-  auto fbrtft = fbrtf.slice(*fbrtf->type_nested_root(), *fbrtf->type());
-  auto fbrtft2 = roundtrip(fbrtft);
-  CHECK_EQUAL(as_bytes(fbrtft.chunk()), as_bytes(fbrtft2.chunk()));
-}
```
Comment (dominiklohmann) [inline libtenzir/test/flatbuffer.cpp:82]: I think we can bring this test back, it used the actor system fixture, but I think it didn't actually need the fixture.

File: nix/caf/source.json [hunk 6]
```diff
@@ -1,7 +1,7 @@
 {
-  "owner": "tenzir",
+  "owner": "actor-framework",
   "repo": "actor-framework",
-  "rev": "4ff66bf218dec0514829230af6f41fb8ed09aa5c",
-  "sha256": "s7CdmQ64whJ7o7KiIxKLrZSR4RRJnNzcgGuqGKcxCzI=",
-  "version": "0.18.7"
+  "rev": "1.0.2",
+  "hash": "sha256-1DJ8VYBTC4Kd6IQZoj4AjP3CoHhb+bmtBEozc5T0R/0=",
```
Comment (dominiklohmann) [inline nix/caf/source.json:5]: Just for my understanding: What's the difference between specifying `sha256` and `hash` for Nix here?

File: tenzir/tenzir.cpp [hunk 7]
```diff
@@ -132,10 +133,13 @@ auto main(int argc, char** argv) -> int {
...
 #endif
-  // Print the configuration file(s) that were loaded.
-  if (!cfg.config_file_path.empty()) {
-    cfg.config_files.emplace_back(std::move(cfg.config_file_path));
+  // Copy CAF detected default config file paths.
+  for (const auto& path : cfg.config_file_paths()) {
+    cfg.config_files.emplace_back(path);
   }
+  // Clear the CAF based default config file paths to avoid duplicates.
+  cfg.config_file_paths({});
```
Comment (dominiklohmann) [inline tenzir/tenzir.cpp:141]: Should this be done in the constructor of `default_configuration` instead? Or maybe I don't understand this entirely, it's a bit hard to follow this.

**Review by dominiklohmann — COMMENTED** [review]
This is fantastic. I started benchmarking this locally and the performance delta appears to be negligible. I read through all the code once now and will do further testing locally, but it looks like we're finally back to upstream CAF soon! ❤️

File: libtenzir/include/tenzir/concept/support/detail/variant.hpp [hunk 8]
```diff
@@ -15,6 +15,24 @@
...
+struct is_variant<std::variant<Args...>> : std::true_type {};
+
+template <typename... Args>
+struct is_variant<const std::variant<Args...>> : std::true_type {};
+
+template <typename... Args>
+struct is_variant<std::variant<Args...>&> : std::true_type {};
+
+template <typename... Args>
+struct is_variant<const std::variant<Args...>&> : std::true_type {};
```
[the remark this answers is not in the document]
Comment (tobim) [reply libtenzir/include/tenzir/concept/support/detail/variant.hpp:31]: I don't think that would work by itself, you would have to add an `same_as` on top. I believe multiple overloads are significantly easier to follow.

File: CMakeLists.txt [hunk 9]
```diff
@@ -304,6 +304,9 @@ elseif (CMAKE_CXX_COMPILER_ID MATCHES "GNU")
   target_compile_options(libtenzir_internal INTERFACE -Wno-unknown-warning)
 endif ()
 
+# FIXME: remove and fix warnings.
```
[the remark this answers is not in the document]
Comment (tobim) [reply CMakeLists.txt:307]: Honestly I didn't really look at them yet.

File: libtenzir/src/configuration.cpp [hunk 10]
```diff
@@ -543,11 +544,12 @@ auto configuration::parse(int argc, char** argv) -> caf::error {
...
-  openssl_key = caf::get_or(content, "caf.openssl.key", "");
-  openssl_passphrase = caf::get_or(content, "caf.openssl.passphrase", "");
-  openssl_capath = caf::get_or(content, "caf.openssl.capath", "");
-  openssl_cafile = caf::get_or(content, "caf.openssl.cafile", "");
+  // FIXME: Are these handled automatically now?
+  // openssl_certificate = caf::get_or(content, "caf.openssl.certificate", "");
+  // openssl_key = caf::get_or(content, "caf.openssl.key", "");
+  // openssl_passphrase = caf::get_or(content, "caf.openssl.passphrase", "");
+  // openssl_capath = caf::get_or(content, "caf.openssl.capath", "");
+  // openssl_cafile = caf::get_or(content, "caf.openssl.cafile", "");
```
[the remark this answers is not in the document]
Comment (tobim) [reply libtenzir/src/configuration.cpp:552]: I think we simply need a tests for encrypted and signed caf IPC.

File: libtenzir/include/tenzir/detail/type_traits.hpp [hunk 11]
```diff
@@ -77,6 +77,13 @@ using is_same_or_derived_t = typename is_same_or_derived<A, B>::type;
 template <class A, class B>
 inline constexpr bool is_same_or_derived_v = is_same_or_derived<A, B>::value;
 
+/// Checks whether `Tpl` is a specialization of `T` or not.
```
[the remark this answers is not in the document]
Comment (tobim) [reply libtenzir/include/tenzir/detail/type_traits.hpp:80]: I think I took that from P2098. Which is actually not in C++23...

File: libtenzir/src/connector.cpp [hunk 12]
```diff
@@ -104,6 +98,18 @@ bool is_recoverable_error_enum(caf::sec err_enum) {
     case caf::sec::broken_promise:
     case caf::sec::connection_timeout:
     case caf::sec::action_reschedule_failed:
+    case caf::sec::invalid_observable:
+    case caf::sec::too_many_observers:
+    case caf::sec::disposed:
+    case caf::sec::cannot_open_resource:
+    case caf::sec::protocol_error:
+    case caf::sec::logic_error:
```
[the remark this answers is not in the document]
Comment (tobim) [reply libtenzir/src/connector.cpp:106]: Yeah, I didn't question the previous logic too much, I just kept the whitelist for the same codes we had before.

File: libtenzir/src/configuration.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (dominiklohmann) [reply libtenzir/src/configuration.cpp:552]: And documentation, too. I never really got this to work for me locally.

File: nix/caf/source.json (hunk 6 above)
Comment (tobim) [reply nix/caf/source.json:5]: This is simply a migration that the nix ecosystem has been doing. Functionally the two arguments are equivalent.

Technically: `hash` takes an sri encoded hash, which specifies the function as part of the string. That way we don't need varargs to support all the different hash functions in wrappers that ultimately boil down to calls to `builtins.derivation` any more.

Comment (dominiklohmann) [reply nix/caf/source.json:5]: Thanks for the explanation 🙏

File: libtenzir/src/default_configuration.cpp (hunk 2 above)
Comment (tobim) [reply libtenzir/src/default_configuration.cpp]: I did some local testing and couldn't measure an effect of `scheduler.max-throughput` any more, so I removed it.

File: tenzir/tenzir.cpp (hunk 7 above)
Comment (tobim) [reply tenzir/tenzir.cpp:141]: This could also be done in our overridden `configuration::parse()`, but it's also not really necessary to refactor that in a big mechanical PR like this one.

File: libtenzir/src/configuration.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (dominiklohmann) [reply libtenzir/src/configuration.cpp:552]: I created https://github.com/tenzir/issues/issues/2381 as a follow-up

**Review by dominiklohmann — APPROVED** [review]
<3
